---
title: "UMCCR Patient Summary"
author: "Oliver Hofmann"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: readable
    toc: false
    toc_float: false
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  workdir: '/Users/vsaveliev/tmp'
  vcf_fname: 'ensemble-with_chr_prefix.vcf'
  af_freqs: 'af/af_tumor.txt'
  af_freqs_az300: 'af/af_tumor_az300.txt'
  sv_fname: 'sv-prioritize-manta-pass.tsv'
  cnvkit_calls: 'cnvkit-call.cns'
  manta_vcf: 'manta.vcf'
  tumor_name: '17MHP031Tmr'
  sig_probs: '/Users/vsaveliev/git/umccr/umccrise/umccrise/patient_analysis/rmd_files/signatures_probabilities.txt'
  suppressors: '/Users/vsaveliev/git/umccr/umccrise/umccrise/patient_analysis/rmd_files/suppressors.txt'
  genome_build: hg19
---

```{r custom, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
library(knitr)
library(rmarkdown)
library(tidyr)
library(dplyr)
library(readr)
library(DT)
library(MutationalPatterns)
library(BSgenome)
library(devtools)
library(stringr)
library(rock)
ref_genome <- paste0("BSgenome.Hsapiens.UCSC.", params$genome_build)
library(ref_genome, character.only = TRUE)
tx_ref_genome <- paste0("TxDb.Hsapiens.UCSC.", params$genome_build, ".knownGene")
library(tx_ref_genome, character.only = TRUE)
library(vcfR)
```

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = params$workdir)
options(width = 140)
```

```{r render_interactively, eval=FALSE, echo=FALSE}
renderMyDocument <- function() {
  rmarkdown::render("sig.Rmd", params = list(
    tumor_name = 'sample_foo',
    vcf_fname = '../../nogit/data/somatic/ensemble-with_chr_prefix.vcf',
    af_freqs = '../../nogit/data/af/af_vardict_tumor.txt',
    af_freqs_az300 = '../../nogit/data/af/af_vardict_tumor_az300.txt',
    sv_fname = '../../nogit/data/structural/sv-prioritize-manta-bpi.tsv',
    sig_probs = 'rmd_files/signatures_probabilities.txt',
    suppressors = 'rmd_files/suppressors.txt',
    genome_build = 'hg19'
  ))
}

# can run this interactively to knit report
renderMyDocument()
```

## Allelic frequencies

Comparing the allelic frequencies of all somatic mutations vs those present in a subset of ~300 known cancer genes. Frequencies are currently based on Mutect2 calls only and limited to 'high confidence' regions as determined by the [Genome in a Botte consortium].

```{r af}
# Global AF
print("params$af_freqs:")
print(params$af_freqs)
print("")
print("params$af_freqs_az300:")
print(params$af_freqs_az300)
print("")

af <- readr::read_tsv(params$af_freqs, col_names = "af", col_types = "d") %>%
  dplyr::mutate(set = "wgs")

# AF for AstraZeneca's AZ300 gene set
af_az300 <- readr::read_tsv(params$af_freqs_az300, col_names = "af", col_types = "-----d-") %>%
  dplyr::mutate(set = 'cancer_genes')

dplyr::bind_rows(af, af_az300) %>%
  ggplot(aes(af)) +
  geom_histogram(stat = 'bin', binwidth = 0.01, fill = "firebrick") +
  facet_wrap(~set, scales = 'free_y') +
  theme_bw()
```

```{r modecalc}
mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}
```
Based on all variants we are seeing a mean allelic frequency of `r round(mean(af$af), 2)` (median `r median(af$af)`, mode `r mode(af$af)`). Looking at just SNVs in known cancer genes this changes to mean `r round(mean(af_az300$af), 2)` (median `r median(af_az300$af)`, mode `r mode(af_az300$af)`).


## Mutational signature

[PCGR](https://github.com/sigven/pcgr) uses [deconstructSigs](https://cran.r-project.org/web/packages/deconstructSigs/index.html) to generate somatic signatures. From our experience it can miss signatures quite frequently; using [MutationalPatterns](http://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html) to doublecheck.

```{r importVCF}
# Grab ensemble calls from the post-processing stage. These were converted
# by adding "chr" prefixes and remiving G* chromosomes.
#

vcf <- read_vcfs_as_granges(
  params$vcf_fname,
  sample_names = params$tumor_name,
  genome = ref_genome,
  group = "auto+sex")
```

### Look at somatic change distribution:

```{r somProfile}
type_occurrences <- mut_type_occurrences(vcf, ref_genome = ref_genome)
plot_spectrum(type_occurrences, CT = TRUE)

# Could do this relative to a few reference samples
mut_mat <- mut_matrix(vcf_list = vcf, ref_genome = ref_genome)
plot_96_profile(mut_mat)
```

### Extract somatic signatures and compare to the reference set:

```{r somSig}
# Get Sanger sigs
# sp_url <- paste("http://cancer.sanger.ac.uk/cancergenome/assets/", "signatures_probabilities.txt", sep = "")
sp_url <- params$sig_probs
# better be explicit - the sig_probs file has 7 extra empty columns
col_types <- paste0(c("ccc", paste0(rep("d", 30), collapse = ""), "ccccccc"), collapse = "")
col_names <- c("SubstType", "Trinucleotide", "SomMutType", paste0("Sig", 1:30), paste0("foo", 1:7))
cancer_signatures <- readr::read_tsv(sp_url, col_names = col_names, col_types = col_types, skip = 1) %>%
  dplyr::arrange(SubstType) %>%
  dplyr::select(4:33) %>%
  as.matrix()

# Fit mutation matrix to cancer signatures. This function finds the optimal linear
# combination of mutation signatures that most closely reconstructs the
# mutation matrix by solving a non-negative least-squares constraints problem.
fit_res <- fit_to_signatures(mut_mat, cancer_signatures)
# Select signatures with some contribution
fit_res_contr <- fit_res$contribution[fit_res$contribution[, 1] > 0, ]
result <- data.frame(Signature = names(fit_res_contr), Contribution = fit_res_contr)

# Quick summary of the results; no plotting yet
result %>%
  dplyr::select(Signature, Contribution) %>%
  dplyr::arrange(-Contribution) %>%
  DT::datatable(rownames = FALSE) %>%
  DT::formatRound('Contribution', 1)
```

### Check for positional enrichment of somatic signatures (limited to autosomes):
```{r rainfall}
chromosomes <- seqnames(get(ref_genome))[1:22]
plot_rainfall(vcf[[1]], chromosomes = chromosomes, cex = 1.5, ylim = 1e+09 )
```

## Prioritised structural variants

```{r sv_prioritize}
col_types <- "ccciicccccc"
sv_all = NULL
if (length(readLines(params$sv_fname, 2)) > 1) {
  sv_all <- readr::read_tsv(params$sv_fname, col_names = TRUE, col_types = col_types) %>%
    tidyr::unnest(annotation = strsplit(annotation, ',')) %>% # Unpack multiple annotations per region
    tidyr::separate(annotation,
                    c('type', 'annotation', 'gene', 'transcript', 'priority', 'score'),
                    sep = '\\|') %>% # Unpack annotation columns %>%
    dplyr::select(chrom, start, end, gene, priority, score, annotation, type, lof,
                  split_read_support, paired_support_PR) %>%
    dplyr::distinct()
  DT::datatable(sv_all, rownames = FALSE)

} else {
  warning('No prioritized events detected')
}
```

### Loss of function in tumor suppressors

```{r sv_suppressors}
if (!is.null(sv_all)) {
  suppressors <- readr::read_tsv(params$suppressors, col_names = "gene", col_types = "c") %>%
    dplyr::pull(gene)

  lof <- sv_all %>%
    dplyr::mutate(lof = stringr::str_replace_all(lof, '\\(|\\)', '')) %>%
    tidyr::unnest(lof = strsplit(lof, ',')) %>%
    tidyr::separate(lof,
             c('lof_gene', 'geneid', 'transcript_count', 'percent_affected'),
             sep = '\\|') %>%
    dplyr::select(chrom, start, end, lof_gene, transcript_count, percent_affected,
           split_read_support, paired_support_PR) %>%
    dplyr::filter(lof_gene %in% suppressors) %>%
    dplyr::distinct()
  DT::datatable(lof, rownames = FALSE)

} else {
  warning('No prioritized events detected')
}
```

## Circos plot showing Manta SVs and CNVkit CNVs

* For the internal lines:
    - The interchromosomal links take the chromosome colour of mate1 of each breakend pair.
    - The intrachromosomal lines are coloured according to the variant type:
        - Deletions: Red
        - Duplications: Green
        - Insertions: Purple
        - Inversions: Orange

```{r circos-manta-cnvkit, fig.width=7, fig.height=7}
sv <- prep_manta_vcf(params$manta_vcf)
cnv <- prep_cnvkit_seg(params$cnvkit_calls)

err <- NULL
tryCatch( { plot_circos(sv = sv, cnv = cnv) }
          , error = function(e) {an.error <<- e})
if (!is.null(err)) {
  warning('Error in circos plot')
  warning(err)
}
```



## Transcriptional strand bias analysis
We can determine if a gene mutation is on the transcribed or non-transcribed
strand, which can be used to evaluate the involvement of transcription-coupled
repair. By convention base substitutions are regarded as C>X or T>X, so we try
to determine whether the C or T base is on the same strand as the gene
definition. Base substitutions on the same strand as the gene definition are
considered 'untranscribed', and on the opposite strand 'transcribed', since the
gene definitions report the coding or sense strand, which is untranscribed. No
strand information is reported for base substitutions that overlap with more
than one gene on different strands.

```{r tran_strand_bias}
# Get known genes table from UCSC
if (params$genome_build == 'hg19') {
  genes_list <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
} else {
  genes_list <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
}

# Mutation count matrix with strand info (4*6*4=96 -> 96*2=192)
mut_mat_s <- mut_matrix_stranded(vcf,
                                 ref_genome = ref_genome,
                                 ranges = genes_list,
                                 mode = "transcription")

# Mutation count per type and strand
strand_counts <- strand_occurrences(mut_mat_s, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias <- strand_bias_test(strand_counts)

# mutation spectrum with strand distinction
plot_strand(strand_counts, mode = "relative")
# effect size of strand bias
plot_strand_bias(strand_bias)
```


## Replicative strand bias analysis
The involvement of replication-associated mechanisms can be evaluated by
testing for a mutational bias between the leading and lagging strand.
The replication strand is dependent on the locations of replication
origins from which DNA replication is fired.
However, replication timing is dynamic and cell-type specific,
which makes replication strand determination less straightforward than
transcriptional strand bias analysis.
Replication timing profiles can be generated with
Repli-Seq experiments. Once the replication direction is defined,
a strand asymmetry analysis can be performed similarly as the transcription
strand bias analysis.

```{r rep_strand_bias}
repli_file <- system.file("extdata/ReplicationDirectionRegions.bed",
                          package = "MutationalPatterns")
# start/stop contain scientific notation, so need to be doubles
repli_strand <- read_tsv(repli_file, col_names = TRUE, col_types = "cddcc") %>% 
  mutate_if(is.character, as.factor)
repli_strand_granges <- GRanges(
  seqnames = repli_strand$Chr,
  ranges = IRanges(start = repli_strand$Start + 1,
                   end = repli_strand$Stop),
  strand_info = repli_strand$Class)

seqlevelsStyle(repli_strand_granges) <- seqlevelsStyle(base::get(ref_genome))

mut_mat_s_rep <- mut_matrix_stranded(vcf, ref_genome = ref_genome,
                                     ranges = repli_strand_granges,
                                     mode = "replication")
# Mutation count per type and strand
strand_counts_rep <- strand_occurrences(mut_mat_s_rep, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias_rep <- strand_bias_test(strand_counts_rep)

plot_strand(strand_counts_rep, mode = "relative")
plot_strand_bias(strand_bias_rep)
```

## Tumour mutational burden analysis

Tumour mutational burden (TMB) "measures the quantity of mutations found in a tumor". 
Also, it is defined as "a quantitative measure of the total number of mutations per 
coding area of a tumor genome." This type of biomarker is currently under study to 
evaluate whether it may help predict the likelihood a patient's response to 
immunotherapy in a range of advanced cancers. 

Tumors that have higher levels of TMB are believed to express more neoantigens – 
a type of cancer-specific antigen – that may allow for a more robust immune response 
and therefore a more durable response to immunotherapy.

```{r tmb}
vcf <-  read.vcfR("~/Documents/UMCCR/data/vcfs/MH17T001P004-somatic-ensemble-pon_hardfiltered.vcf.gz", verbose =  FALSE) 

#Calculating mutations per megabase
fix <- getFIX(vcf)
vcf_number_rows <- nrow(fix)
mutations_megabase <- round(vcf_number_rows/3200, digits = 2)

#Summarizing annotations for variants in the vcf
ann <- vcfR::extract.info(vcf, "ANN")
region_ann <- sapply(ann, function(x){
  y <- strsplit(x, "\\|")[[1]][2]
})
variant_annotation <- unname(region_ann)

#Creating a nice table output for annotations summary
region_ann_df <- data.frame(table(variant_annotation))
kable(region_ann_df, caption = "Table summarizing all annotations in the vcf and the total number of variants suppporting these annotations", format = "markdown")

#Calculating mutations per megabase in coding region
coding_variants = 0
coding_variants <- region_ann %in% c("frameshift_variant", "missense_variant", "missense_variant&splice_region_variant")
coding_variants <- table(coding_variants)
mutations_megabase_coding <- round(as.vector(coding_variants[2])/40, digits = 2) 
#40MB is the estimated size of coding region in human genome - as used by PCGR as well. 
#We can use 36MB if we go with exact calculations, as only 1.2% of the total genome is considered coding. 
#total genome * percent protein coding = 3,000,000,000 * 0.012 = 36,000,000 ~36MB
```

* The _total number of mutations_ in the vcf are **`r vcf_number_rows`** and 
* Number of mutations per megabase are **`r mutations_megabase`**.
* The _total number of mutations in the coding region_ are **`r as.vector(coding_variants[2])`**
* Number of mutations per megabase in the coding region are **`r mutations_megabase_coding`**

## Addendum

```{r sessioninfo}
devtools::session_info()
```

