---
title: "UMCCR Patient Summary"
author: "Oliver Hofmann"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: readable
    toc: true
    toc_float: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  workdir: '/path/to/tmp'
  vcf_fname: 'ensemble-with_chr_prefix.vcf'
  af_freqs: 'af/af_tumor.txt'
  af_freqs_az300: 'af/af_tumor_az300.txt'
  sv_fname: 'sv-prioritize-manta-pass.tsv'
  cnvkit_calls: 'cnvkit-call.cns'
  manta_vcf: 'manta.vcf'
  tumor_name: '17MHP031Tmr'
  sig_probs: 'rmd_files/signatures_probabilities.txt'
  suppressors: 'rmd_files/suppressors.txt'
  genome_build: hg19
---

```{r load-pkgs, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
library(knitr)
library(kableExtra)
library(rmarkdown)
library(tidyr)
library(dplyr)
library(readr)
library(DT)
library(MutationalPatterns)
library(BSgenome)
library(devtools)
library(stringr)
library(rock)
ref_genome <- paste0("BSgenome.Hsapiens.UCSC.", params$genome_build)
library(ref_genome, character.only = TRUE)
tx_ref_genome <- paste0("TxDb.Hsapiens.UCSC.", params$genome_build, ".knownGene")
library(tx_ref_genome, character.only = TRUE)
library(vcfR)
```

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = params$workdir)
options(width = 140)
```

```{r render_interactively, eval=FALSE, echo=FALSE}
# helper function to run report interactively, instead of via the CL.
render_me <- function() {
  dd <- "/Users/pdiakumis/Desktop/projects/umccr/umccrise_test_data"
  rmarkdown::render(
    "sig.Rmd", 
    params = list(
      workdir = '/Users/pdiakumis/Desktop/projects/umccr/umccrise/umccrise/rmd_files',
      vcf_fname = file.path(dd, 'gold_standard/bcbio_test_project/work/cup__cup_tissue/rmd/ensemble-with_chr_prefix.vcf'),
      af_freqs = file.path(dd, 'gold_standard/bcbio_test_project/work/cup__cup_tissue/rmd/afs/af_tumor.txt'),
      af_freqs_az300 = file.path(dd, 'gold_standard/bcbio_test_project/work/cup__cup_tissue/rmd/afs/af_tumor_az300.txt'),
      sv_fname = file.path(dd, 'gold_standard/bcbio_test_project/cup__cup_tissue/structural/cup__cup_tissue-sv-prioritize-manta-pass.tsv'),
      cnvkit_calls = file.path(dd, 'data/bcbio_test_project/final/cup_tissue/cup-cnvkit-call.cns'),
      manta_vcf = file.path(dd, 'gold_standard/bcbio_test_project/work/cup__cup_tissue/structural/ribbon/manta.vcf'),
      tumor_name = '17MHP031Tmr',
      sig_probs = 'signatures_probabilities.txt',
      suppressors = '/Users/pdiakumis/Desktop/projects/umccr/vcf_stuff/src/ngs-utils/ngs_utils/reference_data/suppressors.txt',
      genome_build = 'hg19')
  )
}

render_me()
```

## Allelic frequencies
We can compare the allelic frequencies of all detected somatic mutations (Global) 
vs. those present in a subset of 300 known cancer genes (AZ300 panel). 
Frequencies are currently based on MuTect2 
calls only and limited to 'high confidence' regions as determined by the 
[Genome in a Bottle consortium](http://jimb.stanford.edu/giab/).

```{r af_plot}
# Global AF
af_global <- readr::read_tsv(params$af_freqs, col_names = "af", col_types = "d") %>%
  dplyr::mutate(set = "Global")

# AZ300 gene set AF
af_az300 <- readr::read_tsv(params$af_freqs_az300, 
                            col_names = c("chrom", "pos", "id", "ref", "alt", "af", "ann"), 
                            col_types = "cicccdc") %>%
  dplyr::select(af) %>% 
  dplyr::mutate(set = 'AZ300')

af_both <- dplyr::bind_rows(af_global, af_az300)

ggplot(data = af_both, aes(af)) +
  geom_histogram(stat = 'bin', binwidth = 0.01, fill = "#0047ab") +
  facet_wrap(~set, scales = 'free_y') +
  scale_x_continuous(name = "Allele Frequency", 
                     breaks = seq(0, 1, by = 0.1),
                     limits = c(0, 1), expand = c(0, 0)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        panel.grid.minor = element_blank())
```

### AF summary stats

```{r af_stats}
mode2 <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

af_both %>% 
  dplyr::group_by(set) %>% 
  
  dplyr::summarise(n = n(),
                   mean = round(mean(af), 2),
                   median = round(median(af), 2),
                   mode = round(mode2(af), 2)) %>% 
  knitr::kable() %>% 
  kableExtra::kable_styling(full_width = FALSE, position = "left") %>% 
  kableExtra::column_spec(1, bold = TRUE)
```


## Mutational signature
We can use [MutationalPatterns](http://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html)
to generate a mutation signature for our sample.

```{r importVCF, warning=FALSE}
# Grab ensemble calls from the post-processing stage. These were converted
# by adding "chr" prefixes and removing G* chromosomes.
vcf <- read_vcfs_as_granges(
  params$vcf_fname,
  sample_names = params$tumor_name,
  genome = ref_genome,
  group = "auto+sex")
```

### Somatic change distribution

#### Point mutation spectrum
```{r somProfile1, warning=FALSE}
type_occurrences <- mut_type_occurrences(vcf_list = vcf, ref_genome = ref_genome)
plot_spectrum(type_occurrences, CT = TRUE)
```

#### 96 trinucleotide profile
```{r somProfile2, warning=FALSE}
mut_mat <- mut_matrix(vcf_list = vcf, ref_genome = ref_genome)
plot_96_profile(mut_mat, condensed = TRUE)
```

### Compare signature to reference
```{r somSig}
# Get Sanger sigs from "http://cancer.sanger.ac.uk/cancergenome/assets/signatures_probabilities.txt"
sig_probs <- params$sig_probs
# better be explicit - the sig_probs file has 7 extra empty columns
col_types <- paste0(c("ccc", paste0(rep("d", 30), collapse = ""), "ccccccc"), collapse = "")
col_names <- c("SubstType", "Trinucleotide", "SomMutType", paste0("Sig", 1:30), paste0("foo", 1:7))
cancer_signatures <- readr::read_tsv(sig_probs, col_names = col_names, col_types = col_types, skip = 1) %>%
  dplyr::arrange(SubstType) %>%
  dplyr::select(4:33) %>%
  as.matrix()

# Fit mutation matrix to cancer signatures
fit_res <- fit_to_signatures(mut_matrix = mut_mat, signatures = cancer_signatures)

# Select signatures with some contribution
fit_res_contr <- fit_res$contribution[fit_res$contribution[, 1] > 0, ]
result <- dplyr::tibble(Signature = names(fit_res_contr), Contribution = fit_res_contr)

result %>%
  dplyr::arrange(-Contribution) %>%
  DT::datatable(rownames = FALSE,
                options = list(dom = 't')) %>%
  DT::formatRound('Contribution', 1)
```

### Check for positional enrichment of somatic signatures (limited to autosomes):

```{r rainfall}
chromosomes <- seqnames(get(ref_genome))[1:22]
plot_rainfall(vcf[[1]], chromosomes = chromosomes, cex = 1.5, ylim = 1e+09 )
```

## Prioritised SVs

```{r sv_prioritize}
col_types <- "ccciicccccc"
sv_all = NULL
if (length(readLines(params$sv_fname, 2)) > 1) {
  sv_all <- readr::read_tsv(params$sv_fname, col_names = TRUE, col_types = col_types) %>%
    tidyr::unnest(annotation = strsplit(annotation, ',')) %>% # Unpack multiple annotations per region
    tidyr::separate(annotation,
                    c('type', 'annotation', 'gene', 'transcript', 'priority', 'score'),
                    sep = '\\|', convert = TRUE) %>% # Unpack annotation columns %>%
    dplyr::select(chrom, start, end, gene, priority, score, annotation, type, lof,
                  split_read_support, paired_support_PR) %>%
    dplyr::distinct() %>% 
    dplyr::mutate(chrom = factor(chrom, levels = c(1:22, "X", "Y", "MT")))
  
  DT::datatable(sv_all, rownames = FALSE, filter = "top")
  
} else {
  warning('No prioritized events detected')
}
```

### Loss of function in tumor suppressors

```{r sv_suppressors}
if (!is.null(sv_all)) {
  suppressors <- readr::read_tsv(params$suppressors, col_names = "gene", col_types = "c") %>%
    dplyr::pull(gene)
  
  lof <- sv_all %>%
    dplyr::mutate(lof = stringr::str_replace_all(lof, '\\(|\\)', '')) %>%
    tidyr::unnest(lof = strsplit(lof, ',')) %>%
    tidyr::separate(lof,
                    c('lof_gene', 'geneid', 'transcript_count', 'percent_affected'),
                    sep = '\\|') %>%
    dplyr::select(chrom, start, end, lof_gene, transcript_count, percent_affected,
                  split_read_support, paired_support_PR) %>%
    dplyr::filter(lof_gene %in% suppressors) %>%
    dplyr::distinct()
  
  DT::datatable(lof, rownames = FALSE) %>%
    DT::formatCurrency('start', currency = "", interval = 3, mark = ",", digits = 0) %>% 
    DT::formatCurrency('end', currency = "", interval = 3, mark = ",", digits = 0)
  
} else {
  warning('No prioritized events detected')
}
```

## OmicCircos plot (Manta + CNVkit)

* For the internal lines:
    - The interchromosomal links take the chromosome colour of mate1 of each breakend pair.
    - The intrachromosomal lines are coloured according to the variant type:
        - Deletions: Red
        - Duplications: Green
        - Insertions: Purple
        - Inversions: Orange

```{r circos-manta-cnvkit, fig.width=7, fig.height=7, warning=FALSE}
sv <- prep_manta_vcf(params$manta_vcf)
cnv <- prep_cnvkit_seg(params$cnvkit_calls)
plot_circos(sv = sv, cnv = cnv)
```

## Perl Circos plot 

```{r circos-prep, eval=FALSE, message=FALSE, warning=FALSE}
circos_prep(outdir = file.path(params$workdir, "circos"), manta = params$manta_vcf, cnv = params$cnvkit_calls)
plot_circos2(outdir = "circos", name = params$tumor_name)
```

```{r circos-manta-cnvkit2, out.width='80%'}
knitr::include_graphics(file.path(params$workdir, "circos", paste0(params$tumor_name, "_circos_cnvkit_manta.png")))
```


## Transcriptional strand bias analysis

We can determine if a gene mutation is on the transcribed or non-transcribed
strand, which can be used to evaluate the involvement of transcription-coupled
repair. By convention base substitutions are regarded as C>X or T>X, so we try
to determine whether the C or T base is on the same strand as the gene
definition. Base substitutions on the same strand as the gene definition are
considered 'untranscribed', and on the opposite strand 'transcribed', since the
gene definitions report the coding or sense strand, which is untranscribed. No
strand information is reported for base substitutions that overlap with more
than one gene on different strands.

```{r tran_strand_bias, warning=FALSE, message=FALSE}
# Get known genes table from UCSC
if (params$genome_build == 'hg19') {
  genes_list <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
} else {
  genes_list <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
}

# Mutation count matrix with strand info (4*6*4=96 -> 96*2=192)
mut_mat_s <- mut_matrix_stranded(vcf,
                                 ref_genome = ref_genome,
                                 ranges = genes_list,
                                 mode = "transcription")

# Mutation count per type and strand
strand_counts <- strand_occurrences(mut_mat_s, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias <- strand_bias_test(strand_counts)

# mutation spectrum with strand distinction
plot_strand(strand_counts, mode = "relative")
# effect size of strand bias
plot_strand_bias(strand_bias)
```


## Replicative strand bias analysis

The involvement of replication-associated mechanisms can be evaluated by
testing for a mutational bias between the leading and lagging strand.
The replication strand is dependent on the locations of replication
origins from which DNA replication is fired.
However, replication timing is dynamic and cell-type specific,
which makes replication strand determination less straightforward than
transcriptional strand bias analysis.
Replication timing profiles can be generated with
Repli-Seq experiments. Once the replication direction is defined,
a strand asymmetry analysis can be performed similarly as the transcription
strand bias analysis.

```{r rep_strand_bias, warning=FALSE, message=FALSE}
repli_file <- system.file("extdata/ReplicationDirectionRegions.bed",
                          package = "MutationalPatterns")
# start/stop contain scientific notation, so need to be doubles
repli_strand <- read_tsv(repli_file, col_names = TRUE, col_types = "cddcc") %>% 
  mutate_if(is.character, as.factor)
repli_strand_granges <- GRanges(
  seqnames = repli_strand$Chr,
  ranges = IRanges(start = repli_strand$Start + 1,
                   end = repli_strand$Stop),
  strand_info = repli_strand$Class)

seqlevelsStyle(repli_strand_granges) <- seqlevelsStyle(base::get(ref_genome))

mut_mat_s_rep <- mut_matrix_stranded(vcf, ref_genome = ref_genome,
                                     ranges = repli_strand_granges,
                                     mode = "replication")
# Mutation count per type and strand
strand_counts_rep <- strand_occurrences(mut_mat_s_rep, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias_rep <- strand_bias_test(strand_counts_rep)

plot_strand(strand_counts_rep, mode = "relative")
plot_strand_bias(strand_bias_rep)
```

## Tumour mutational burden analysis

Tumour mutational burden (TMB) "measures the quantity of mutations found in a tumor".
Also, it is defined as "a quantitative measure of the total number of mutations per 
coding area of a tumor genome." This type of biomarker is currently under study to 
evaluate whether it may help predict the likelihood a patient's response to 
immunotherapy in a range of advanced cancers. 

Tumors that have higher levels of TMB are believed to express more neoantigens -
a type of cancer-specific antigen - that may allow for a more robust immune response 
and therefore a more durable response to immunotherapy.

```{r tmb}
vcf <-  read.vcfR(params$vcf_fname, verbose =  FALSE) 

#Calculating mutations per megabase
fix <- getFIX(vcf)
vcf_number_rows <- nrow(fix)
mutations_megabase <- round(vcf_number_rows/3200, digits = 2)

#Summarizing annotations for variants in the vcf
ann <- vcfR::extract.info(vcf, "ANN")
region_ann <- sapply(ann, function(x){
  y <- strsplit(x, "\\|")[[1]][2]
})
variant_annotation <- unname(region_ann)

#Creating a nice table output for annotations summary
region_ann_df <- data.frame(table(variant_annotation))
kable(region_ann_df, caption = "Table summarizing all annotations in the vcf and the total number of variants suppporting these annotations", format = "markdown")

#Calculating mutations per megabase in coding region
coding_variants = 0
coding_variants <- region_ann %in% c("frameshift_variant", "missense_variant", "missense_variant&splice_region_variant")
coding_variants <- table(coding_variants)
mutations_megabase_coding <- round(as.vector(coding_variants[2])/40, digits = 2) 
#40MB is the estimated size of coding region in human genome - as used by PCGR as well. 
#We can use 36MB if we go with exact calculations, as only 1.2% of the total genome is considered coding. 
#total genome * percent protein coding = 3,000,000,000 * 0.012 = 36,000,000 ~36MB

#Displaying results in a table
region <- c("Wholegenome", "Coding")
total_mutations <- c(vcf_number_rows, as.vector(coding_variants[2]))
mutations_mb <- c(mutations_megabase, mutations_megabase_coding)

result_display <- data.frame(region, total_mutations, mutations_mb) 
kable(result_display,  caption = "Table summarizing somatic burden result")
```

## Software Used

<details>
<summary>R packages</summary>
```{r}
si <- devtools::session_info(include_base = TRUE)
si_pl <- unclass(si$platform) %>% as_tibble() %>% t()
si_pkg <- unclass(si$packages) %>% as_tibble()

dplyr::tibble(var = row.names(si_pl), 
              value = si_pl[, , drop = TRUE]) %>% 
  knitr::kable()

knitr::kable(si_pkg)
```
</details>

<details>
<summary>conda packages</summary>

```{r eval=FALSE}
which_r <- Sys.which("R")
conda_bin <- list.files(file.path(dirname(dirname(dirname(dirname(x)))), "bin"), 
                        pattern = "^conda$", full.names=T)
conda_list <- system(command = paste(conda_bin, "list"), intern = TRUE)
conda_list # do pretty stuff with it
```

</details>

